### PR TITLE
Add company and office

### PR DESCRIPTION
**Pull request by BelehovEgor**
http://outsourcing.nat.tepkom.ru:20080/redmine/issues/703

### CODE REVIEW

File: src/CompanyService.Business/Commands/Company/CreateCompanyCommand.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        private Guid? AddImage(AddImageRequest image, List<string> errors)
+        {
+            Guid creatorId = _httpContextAccessor.HttpContext.GetUserId();
+            string logMessage = "Cannot add image: name - {name}, context - {context}, extension - {extension}, creatorId - {creatorId}";
+            string errorMessage = "Cannot add company logo. Please try later.";
+
+            try
+            {
+                var response = _rcAddImage.GetResponse<IOperationResult<IAddImageResponse>>(
+                                                IAddImageRequest.CreateObj(image.Name, image.Content, image.Extension, creatorId)).Result.Message;
```
Comment (MariyaSin) [inline src/CompanyService.Business/Commands/Company/CreateCompanyCommand.cs:44]: remove extra spaces

File: src/CompanyService.Business/Commands/Company/GetCompanyCommand.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+        private ImageInfo GetImage(Guid imageId, List<string> errors)
+        {
+            string logMessage = "Cannot get image with id: {id}";
+            string errorMessage = $"Cannot get image with id: {imageId}. Please try later.";
+
+            try
+            {
+                var response = _rcGetFile.GetResponse<IOperationResult<IGetFileResponse>>(
+                                                IGetFileRequest.CreateObj(imageId)).Result.Message;
```
Comment (MariyaSin) [inline src/CompanyService.Business/Commands/Company/GetCompanyCommand.cs:32]: remove extra spaces

**Review by MariyaSin — APPROVED** [review]

File: src/CompanyService.Business/Commands/Company/CreateCompanyCommand.cs (hunk 1 above)
Comment (BelehovEgor) [reply src/CompanyService.Business/Commands/Company/CreateCompanyCommand.cs:44]: I did this on purpose so that the method arguments are underneath it.

Comment (BelehovEgor) [reply src/CompanyService.Business/Commands/Company/CreateCompanyCommand.cs]: you are rigth, fixed

File: src/CompanyService.Business/Commands/Company/GetCompanyCommand.cs (hunk 2 above)
Comment (BelehovEgor) [reply src/CompanyService.Business/Commands/Company/GetCompanyCommand.cs]: fixed

**Review by borovHokage — APPROVED** [review]

File: src/CompanyService.Validation/CreateCompanyRequestValidator.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    public class CreateCompanyRequestValidator : AbstractValidator<CreateCompanyRequest>,  ICreateCompanyRequestValidator
+    {
+        public CreateCompanyRequestValidator()
+        {
+            RuleFor(request => request.Name)
+                .NotEmpty();
+
+            RuleFor(request => request.SiteUrl)
+                .NotEmpty();
+        }
```
Comment (djino97) [inline src/CompanyService.Validation/CreateCompanyRequestValidator.cs:16]: Where tagline validation?

File: src/CompanyService.Validation/CreateOfficeRequestValidator.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+using LT.DigitalOffice.CompanyService.Models.Dto.Requests;
+using LT.DigitalOffice.CompanyService.Validation.Interfaces;
+
+namespace LT.DigitalOffice.CompanyService.Validation
+{
+    public class CreateOfficeRequestValidator : AbstractValidator<CreateOfficeRequest>, ICreateOfficeRequestValidator
+    {
+        public CreateOfficeRequestValidator()
+        {
+            RuleFor(request => request.City)
```
Comment (djino97) [inline src/CompanyService.Validation/CreateOfficeRequestValidator.cs:12]: Where name validation?

File: src/CompanyService/Controllers/OfficeController.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        public OperationResultResponse<Guid> Create(
+            [FromServices] ICreateOfficeCommand command,
+            [FromBody] CreateOfficeRequest request)
+        {
+            return command.Execute(request);
+        }
+
+        [HttpGet("find")]
+        public OfficesResponse Find(
+            [FromServices] IFindOfficesCommand command)
```
Comment (djino97) [inline src/CompanyService/Controllers/OfficeController.cs:24]: add skipCount and takeCount

File: src/CompanyService.Data/CompanyRepository.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+            if (_provider.Companies.Any())
+            {
+                throw new BadRequestException("Company already exist");
+            }
+
+            _provider.Companies.Add(company);
+            _provider.Save();
+        }
+
+        public DbCompany Get(bool full)
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/CompanyService.Data/CompanyRepository.cs:37]: If full == true uses includes

File: src/CompanyService.Business/Commands/Company/CreateCompanyCommand.cs [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                if (response.IsSuccess)
+                {
+                    return response.Body.Id;
+                }
+
+                _logger.LogWarning(logMessage, image.Name, image.Content, image.Extension, creatorId);
+            }
+            catch (Exception exc)
+            {
+                _logger.LogWarning(exc, logMessage, image.Name, image.Content, image.Extension, creatorId);
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/CompanyService.Business/Commands/Company/CreateCompanyCommand.cs:55]: fixed

File: src/CompanyService.Business/Commands/Company/GetCompanyCommand.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                if (response.IsSuccess)
+                {
+                    return _imageInfoMapper.Map(response.Body);
+                }
+
+                _logger.LogWarning(logMessage, imageId);
+            }
+            catch (Exception exc)
+            {
+                _logger.LogWarning(exc, logMessage, imageId);
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/CompanyService.Business/Commands/Company/GetCompanyCommand.cs:44]: fixed

File: src/CompanyService.Business/Commands/Office/CreateOfficeCommand.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            }
+
+            _validator.ValidateAndThrowCustom(request);
+
+            DbOffice office = _mapper.Map(request, _companyRepository.Get(false).Id);
+            _officeRepository.Add(office);
+
+            return new OperationResultResponse<Guid>
+            {
+                Status = Kernel.Enums.OperationResultStatusType.FullSuccess,
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/CompanyService.Business/Commands/Office/CreateOfficeCommand.cs:51]: fixed

File: src/CompanyService/Controllers/OfficeController.cs (hunk 5 above)
Comment (BelehovEgor) [reply src/CompanyService/Controllers/OfficeController.cs:24]: fixed

File: src/CompanyService.Data/Interfaces/IOfficeRepository.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
+﻿using LT.DigitalOffice.CompanyService.Models.Db;
+using LT.DigitalOffice.Kernel.Attributes;
+using System.Collections.Generic;
+
+namespace LT.DigitalOffice.CompanyService.Data.Interfaces
+{
+    [AutoInject]
+    public interface IOfficeRepository
+    {
+        void Add(DbOffice office);
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/CompanyService.Data/Interfaces/IOfficeRepository.cs:10]: fixed

File: src/CompanyService.Business/Commands/Company/CreateCompanyCommand.cs [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+            }
+            catch(Exception exc)
+            {
+                _logger.LogError(exc, message);
+
+                throw;
+            }
+            _logger.LogError(message);
+
+            throw new InternalServerException(message);
```
Comment (djino97) [inline src/CompanyService.Business/Commands/Company/CreateCompanyCommand.cs:52]: this is  bad , we not throw server exception our clients. We not stop service working, but return error message

File: src/CompanyService.Business/Commands/Company/CreateCompanyCommand.cs [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+                if (response.IsSuccess && response.Body)
+                {
+                    return;
+                }
+            }
+            catch(Exception exc)
+            {
+                _logger.LogError(exc, message);
+
+                throw;
```
Comment (djino97) [inline src/CompanyService.Business/Commands/Company/CreateCompanyCommand.cs:48]: add error message to error list

**Review by djino97 — APPROVED** [review]

File: src/CompanyService.Business/Commands/Company/GetCompanyCommand.cs [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+        public GetCompanyCommand(
+            ICompanyRepository repository,
+            ICompanyInfoMapper mapper)
+        {
+            _repository = repository;
+            _companyInfoMapper = mapper;
+        }
+
+        public CompanyInfo Execute()
```
[the remark this answers is not in the document]
Comment (BelehovEgor) [reply src/CompanyService.Business/Commands/Company/GetCompanyCommand.cs]: fixed

**Review by MariyaSin — APPROVED** [review]

**Pull request merged**